File: algorithms/238.py
# https://leetcode.com/problems/product-of-array-except-self


class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        len_n = len(nums)
        result = [1] * len_n

        # left --> right
        prod = 1
        for i in range(1, len_n):
            prod *= nums[i - 1]
            result[i] *= prod

        # left <-- right
        prod = 1
        for i in range(len_n - 2, -1, -1):
            prod *= nums[i + 1]
            result[i] *= prod

        return result

File: algorithms/238.swift
// https://leetcode.com/problems/product-of-array-except-self

class Solution {
    func productExceptSelf(nums: [Int]) -> [Int] {
        var result = [Int](count: nums.count, repeatedValue:1)
        var i = 0
        var j = nums.count-1
        var left = 1
        var right = 1

        while j > 0 {
            left *= nums[i]
            right *= nums[j]
            result[i+1] *= left
            result[j-1] *= right
            i += 1
            j -= 1
        }

        return result
    }
}
